File: go/state_of_tic_tac_toe.go
package stateoftictactoe

import (
	"errors"
)

type State string

const (
	Win     State = "win"
	Ongoing State = "ongoing"
	Draw    State = "draw"
)

func StateOfTicTacToe(board []string) (State, error) {
	xs, os := 0, 0
	finished := true
	for _, row := range board {
		for _, c := range row {
			if c == 'X' {
				xs++
			} else if c == 'O' {
				os++
			} else {
				finished = false
			}
		}
	}

	diff := xs - os
	if !(diff >= 0 && diff <= 1) {
		return "", errors.New("invalid")
	}

	rules := [][][]int{
		{{0, 0}, {0, 1}, {0, 2}},
		{{1, 0}, {1, 1}, {1, 2}},
		{{2, 0}, {2, 1}, {2, 2}},
		{{0, 0}, {1, 0}, {2, 0}},
		{{0, 1}, {1, 1}, {2, 1}},
		{{0, 2}, {1, 2}, {2, 2}},
		{{0, 0}, {1, 1}, {2, 2}},
		{{0, 2}, {1, 1}, {2, 0}},
	}

	xwin := false
	owin := false

	for _, rule := range rules {
		mp := map[byte]int{}
		for _, pos := range rule {
			mp[board[pos[0]][pos[1]]]++
		}
		if mp['X'] == 3 {
			xwin = true
		} else if mp['O'] == 3 {
			owin = true
		}
	}

	if xwin && owin {
		return "", errors.New("invalid")
	}

	if xwin || owin {
		return Win, nil
	}

	if !finished {
		return Ongoing, nil
	}

	return Draw, nil
}

File: elixir/state_of_tic_tac_toe.ex
defmodule StateOfTicTacToe do
  @doc """
  Determine the state a game of tic-tac-toe where X starts.
  """
  @spec game_state(board :: String.t()) :: {:ok, :win | :ongoing | :draw} | {:error, String.t()}
  def game_state(board) do
    board =
      board
      |> String.split("\n")
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {r, i}, acc ->
        r
        |> String.graphemes()
        |> Enum.with_index()
        |> Enum.reduce(acc, fn {c, j}, acc ->
          Map.put(acc, {i, j}, c)
        end)
      end)

    xs = Enum.count(board, fn {_, v} -> v == "X" end)
    os = Enum.count(board, fn {_, v} -> v == "O" end)
    x? = check_winner(board, "X")
    o? = check_winner(board, "O")

    cond do
      x? and o? -> {:error, "Impossible board: game should have ended after the game was won"}
      os > xs -> {:error, "Wrong turn order: O started"}
      xs - os > 1 -> {:error, "Wrong turn order: X went twice"}
      x? or o? -> {:ok, :win}
      xs + os == 9 -> {:ok, :draw}
      true -> {:ok, :ongoing}
    end
  end

  defp check_winner(board, t) do
    Enum.any?(0..2, fn i ->
      Enum.all?(0..2, fn j -> board[{i, j}] == t end)
    end) or
      Enum.any?(0..2, fn j ->
        Enum.all?(0..2, fn i -> board[{i, j}] == t end)
      end) or
      Enum.all?(0..2, fn i -> board[{i, i}] == t end) or
      Enum.all?(0..2, fn i -> board[{i, 2 - i}] == t end)
  end
end
